server/src/main/java/org/candlepin/auth/KeycloakAuth.java

```diff
@@ -104,20 +119,28 @@ public Principal getPrincipal(HttpRequest httpRequest) {
                     return null;
                 }
                 else {
-                    Base64 base64Url = new Base64(true);
-                    String[] splitstring = authArray[1].split("\\.");
-                    String base64EncodedBody = splitstring[1];
-                    String body = new String(base64Url.decode(base64EncodedBody));
-                    JSONObject obj = new JSONObject(body);
-                    String tokenType = obj.getString("typ");
-
-                    if (tokenType.equals("Refresh")) {
-                        //Offline Token authentication
-                        handleRefreshToken(httpRequest, auth);
+                    try {
+                        Base64 base64Url = new Base64(true);
+                        String[] splitstring = authArray[1].split("\\.");
+                        String base64EncodedBody = splitstring[1];
+                        String body = new String(base64Url.decode(base64EncodedBody));
+                        ObjectMapper objectMapper = new ObjectMapper();
+                        JsonNode rootNode = objectMapper.readTree(body);
+                        String tokenType = rootNode.get("typ").asText();
+                        Token token = Token.valueOf(tokenType.toUpperCase());
+                        switch(token) {
+                            case BEARER:
+                                handleBearerToken(httpRequest);
+                                break;
+                            case REFRESH:
+                                handleRefreshToken(httpRequest, auth);
+                                break;
+                            default:
+                                break;
+                        }
                     }
-                    else if (tokenType.equals("Bearer")) {
-                        //Bearer only authentication
-                        handleBearerToken(httpRequest);
+                    catch (UnsupportedEncodingException e) {
+                        throw new UnsupportedEncodingException(i18nProvider.get().tr("Decoding Failed"));
```

Comment:
In a case like this where we're throwing the exact same exception that we just caught, we're better off not catching it at all. In this particular case, because the new exception doesn't retain the original as its cause, we lose the origin of the problem and all of the stack information with it.